packages/marshal/src/marshal.js

```diff
@@ -176,68 +256,111 @@ export const makeMarshal = (
+        assert.fail(X`invalid error name typeof ${q(typeof dName)}`);
+      typeof dMessage === 'string' ||
+        assert.fail(X`invalid error message typeof ${q(typeof dMessage)}`);
+      const EC = getErrorConstructor(dName) || Error;
       // errorId is a late addition so be tolerant of its absence.
       const errorName =
-        errorId === undefined
+        dErrorId === undefined
           ? `Remote${EC.name}`
-          : `Remote${EC.name}(${errorId})`;
-      // Due to a defect in the SES type definition, the next line is
-      // fails a type check.
-      // Pending https://github.com/endojs/endo/issues/977
-      // @ts-ignore-next-line
-      const error = assert.error(`${message}`, EC, { errorName });
-      return error;
+          : `Remote${EC.name}(${dErrorId})`;
+      const error = assert.error(dMessage, EC, { errorName });
+      return harden(error);
     };
 
     // The current encoding does not give the decoder enough into to distinguish
     // whether a slot represents a promise or a remotable. As an implementation
     // restriction until this is fixed, if either is provided, both must be
     // provided and they must be the same.
     // See https://github.com/Agoric/agoric-sdk/issues/4334
-    const decodeRemotableOrPromiseFromCapData = rawTree => {
-      const { index, iface } = rawTree;
-      const val = unserializeSlot(index, iface);
-      return val;
+    const decodeRemotableOrPromiseFromCapData = (rawTree, _decodeRecur) => {
+      const { [QCLASS]: _, ...slotData } = rawTree;
+      return decodeSlotCommon(slotData);
     };
 
-    const fullRevive = makeDecodeFromCapData({
+    const decodeErrorFromCapData = (rawTree, decodeRecur) => {
+      const { [QCLASS]: _, ...errData } = rawTree;
+      return decodeErrorCommon(errData, decodeRecur);
+    };
+
+    const reviveFromCapData = makeDecodeFromCapData({
       decodeRemotableFromCapData: decodeRemotableOrPromiseFromCapData,
       decodePromiseFromCapData: decodeRemotableOrPromiseFromCapData,
       decodeErrorFromCapData,
     });
-    return fullRevive;
+
+    const makeDecodeSlotFromSmallcaps = prefix => {
+      return (stringEncoding, _decodeRecur) => {
+        assert(stringEncoding.startsWith(prefix));
+        // slots: $slotIndex.iface or $slotIndex
+        const i = stringEncoding.indexOf('.');
+        const index = Number(stringEncoding.slice(1, i < 0 ? undefined : i));
+        // i < 0 means there was no iface included.
+        const iface = i < 0 ? undefined : stringEncoding.slice(i + 1);
+        return decodeSlotCommon({ iface, index });
+      };
+    };
+    const decodeRemotableFromSmallcaps = makeDecodeSlotFromSmallcaps('$');
+    const decodePromiseFromSmallcaps = makeDecodeSlotFromSmallcaps('&');
+
+    const decodeErrorFromSmallcaps = (encoding, decodeRecur) => {
+      const { '#error': message, ...restErrData } = encoding;
+      return decodeErrorFromCapData({ message, ...restErrData }, decodeRecur);
```

Comment:
```suggestion
      !hasOwnPropertyOf(restErrData, 'message') ||
        assert.fail(X`unexpected encoded error property ${q('message')}`);
      return decodeErrorCommon({ message, ...restErrData }, decodeRecur);
```

Reply:
Good catch! I completely missed that `restErrData` might have a `message` property that would then overwrite the message text obtained from `#error`. Done.